commit/report.go

```diff
@@ -98,6 +84,189 @@ func (p *Plugin) Reports(
+		GasPriceUpdates:   outcome.ChainFeeOutcome.GasPrices,
+	}
+
+	report, err := buildOneReport(
+		ctx,
+		lggr,
+		reportCodec,
+		transmission,
+		outcome.MerkleRootOutcome.OutcomeType,
+		blessedMerkleRoots,
+		unblessedMerkleRoots,
+		outcome.MerkleRootOutcome.RMNReportSignatures,
+		outcome.MerkleRootOutcome.RMNRemoteCfg.FSign,
+		priceUpdates,
+	)
+	if err != nil {
+		return nil, err
+	}
+	if report != nil {
+		return []ocr3types.ReportPlus[[]byte]{*report}, nil
+	}
+	return nil, nil
+}
+
+// buildMultipleReports builds many reports of with at most maxMerkleRootsPerReport roots.
+func buildMultipleReports(
+	ctx context.Context,
+	lggr logger.Logger,
+	reportCodec cciptypes.CommitPluginCodec,
+	transmissionSchedule *ocr3types.TransmissionSchedule,
+	outcome committypes.Outcome,
+	maxMerkleRootsPerReport uint64,
+) ([]ocr3types.ReportPlus[[]byte], error) {
+	var reports []ocr3types.ReportPlus[[]byte]
+
+	numRoots := uint64(0)
+	blessedMerkleRoots := make([]cciptypes.MerkleRootChain, 0)
+	unblessedMerkleRoots := make([]cciptypes.MerkleRootChain, 0)
+
+	priceUpdates := cciptypes.PriceUpdates{
+		TokenPriceUpdates: outcome.TokenPriceOutcome.TokenPrices.ToSortedSlice(),
+		GasPriceUpdates:   outcome.ChainFeeOutcome.GasPrices,
+	}
+
+	for _, r := range outcome.MerkleRootOutcome.RootsToReport {
+
+		// TODO: Support RMN.
+		/*
+			if outcome.MerkleRootOutcome.RMNEnabledChains[r.ChainSel] {
+				blessedMerkleRoots = append(blessedMerkleRoots, r)
+			} else {
+				unblessedMerkleRoots = append(unblessedMerkleRoots, r)
+			}
+		*/
+		unblessedMerkleRoots = append(unblessedMerkleRoots, r)
+		numRoots++
+
+		if numRoots == maxMerkleRootsPerReport {
+			report, err := buildOneReport(
+				ctx,
+				lggr,
+				reportCodec,
+				transmissionSchedule,
+				outcome.MerkleRootOutcome.OutcomeType,
+				blessedMerkleRoots,
+				unblessedMerkleRoots,
+				nil, // no RMN for partial reports.
+				0,   // no RMN for partial reports.
+				priceUpdates,
+			)
+			if err != nil {
+				return nil, err
+			}
+			if report != nil {
+				reports = append(reports, *report)
+			}
+
+			// reset accumulators for next report.
+			numRoots = 0
+			blessedMerkleRoots = make([]cciptypes.MerkleRootChain, 0)
+			unblessedMerkleRoots = make([]cciptypes.MerkleRootChain, 0)
+
+			// price updates are only included in the first report.
+			priceUpdates = cciptypes.PriceUpdates{}
+		}
+	}
+
+	// check for final partial report.
+	if numRoots > 0 {
+		report, err := buildOneReport(
+			ctx,
+			lggr,
+			reportCodec,
+			transmissionSchedule,
+			outcome.MerkleRootOutcome.OutcomeType,
+			blessedMerkleRoots,
+			unblessedMerkleRoots,
+			nil, // no RMN for partial reports.
+			0,   // no RMN for partial reports.
+			priceUpdates,
+		)
+		if err != nil {
+			return nil, err
+		}
+		if report != nil {
+			reports = append(reports, *report)
+		}
+	}
+
+	return reports, nil
+}
+
+// buildTruncatedReport returns a single truncated report.
```

Comment:
Its not clear what a truncated report is, we seem to trim the slice, but no reason as to why we're doing that is written far as I can tell

Reply:
It's just more flexibility for the NonEVMs if they need it.

Reply:
I'd rather not add flexibility just for the sake of it because its more code to maintain, understand, test, etc. Do we have test runs where this truncated approach works and the multiple reports doesn't, or vice versa? cc @archseer

Reply:
This is the code we're talking about:
```
	reports, err := buildMultipleReports(ctx, lggr, reportCodec, transmissionSchedule, outcome, maxMerkleRootsPerReport)
	if err != nil {
		return nil, err
	}
	if len(reports) > 1 {
		lggr.Warnf("buildTruncatedReport: truncating report %d -> 1", len(reports))
		return reports[:1], nil
	}

	return reports, nil
```

I  don't understand the hangup here. It was easier to handle the case when `MultiplReportsEnabled == false` and `MaxMerkleRootsPerReport != 0` than it was to decide how to handle the error.

At this point the code is written and tested.

I'm not sure why we've spent so much time discussing this point when the wait for transmission function is apparently broken.